vsintegration/src/FSharp.ProjectSystem.FSharp/ProjectFlagsService.fs

```diff
@@ -0,0 +1,128 @@
+                buildResource.Dispose()
+                project.NotifySourcesAndFlags()
+                accessor.EndDesignTimeBuild() |> ignore
+                completeEvent.Trigger()
+
+            // Notes regarding the _ResolveReferenceDependencies property
+            // TODO: check whether this can be achieved in a less painful way 
+            // (names that start with '_' are treated as implementation details and not recommended
+            // for usage).
+            // Setting this property enforces MSBuild to resolve second order dependencies, so if
+            // desktop applications references .NET Core based portable assemblies then MSBuild
+            // will detect that one of references requires System.Runtime and expand list of
+            // references with framework facades for portables.
+            // Usually facades are located at: %programfiles(x86)%\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\Facades\
+            // If the property is not set then MSBuild will resolve only primary dependencies, and
+            // the compiler will be very unhappy as all fundamental types should be taken from
+            // System.Runtime that is not supplied.
+            
+            let projectInstance = ref null
+            UIThread.Run(fun () ->
+                project.DoMSBuildSubmission(BuildKind.ASYNC, "Compile", projectInstance, MSBuildCoda onComplete, [KeyValuePair("_ResolveReferenceDependencies", "true")]) |> ignore
+                )
+            
+            // Wait until the build is complete before we continue servicing the queue
+            do! Async.AwaitEvent completeEvent.Publish
+        }
+
+        async {
+            while true do
+                use processor = MailboxProcessor.Start <| fun inbox -> async {
+                    // Wait for AfterOpenSolution and then start processing
+                    do! Async.AwaitEvent Events.SolutionEvents.OnAfterOpenSolution |> Async.Ignore
+
+                    while true do
+                        let! project = inbox.Receive()
+
+                        // Set the project to clean and keep track of whether it was dirty
+                        let wasDirty =
+                            lock queueLock <| fun () ->
+                                let oldDirty = dirtyProjects.[project]
+                                dirtyProjects.[project] <- false
+                                oldDirty
+                        
+                        // Only process the project if it was still dirty
+                        if wasDirty then
+                            do! processProject project
+
+                            // Having an exclusive lock on design-time builds means the user
+                            // can't initiate any builds themselves. Wait a few seconds before
+                            // trying to grab one again.
+                            do! Async.Sleep 2000
```

Comment:
It leaves some time so that the user can initiate builds/move files around in the Solution Explorer. When a design-time build is running, quite a few parts of the UI are locked out.